<a href="https://colab.research.google.com/github/MOOwuttichai/BSC_DPDM2023/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
import os

In [3]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC-dpdm23_flie'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [5]:
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

        InvoiceDate  UnitPrice  CustomerID         Country  
0  12/01/2010 08:26       2.55     17850.0  United Kingdom  
1  12/01/2010 08:26       3.39     17850.0  United Kingdom  
2  12/01/2010 08:26       2.75     17850.0  United Kingdom  
3  12/01/2010 08:26       3.39     17850.0  United Kingdom  
4  12/01/2010 08:26       3.39     17850.0  United Kingdom

#Quit1 มีกี่ประเทศ

In [6]:
len(set(data['Country']))

38

#เเปลง data ใก้อยู่ในรูปเเบบ Transaction data ด้วย ```groupby.apply(list) ```

In [7]:
#data.groupby('InvoiceNo')['StockCode'].apply(list)

In [8]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [9]:
!pip install apyori

In [10]:
from apyori import apriori

In [11]:
!pip install apyori
from apyori import apriori
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [12]:
len(frequentItemsets)

4

In [13]:
frequentItemsets[0]

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [14]:
frequentItemsets[0][0]

frozenset({'22423'})

In [15]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [16]:
frequentItemsets[0][0].__len__()#หาขนาดของ itemset

1

In [17]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),
support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [18]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

#HW 1 หา Frequent Item Set เเละ Association Rules
* ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)
* ของช่วงเวลา Christmas & New Year ในทวีป Europe(1 Dec - 31 Dec)
* ของวันวาเลนไทน์ (14 Feb)
* จัดกลุ่มตามประเทศ+เดิอน

##1.ของช่วงเวลา Christmas & New Year (1 Dec - 31 Dec)

###เเยกวันเดือนปีออกจากเวลา

In [19]:
data['InvoiceDate'][0].split()[0]

'12/01/2010'

In [20]:
date =[]
for i in range(len(data['InvoiceDate'])):
  x = data['InvoiceDate'][i].split()[0]
  date.append(x)
#set(L)

In [21]:
date[0].split('/')

['12', '01', '2010']

###เเยกวัน เดือน ปีออกจากกกัน

In [22]:
day =[]
month =[]
year =[]
for i in range(len(data['InvoiceDate'])):
  x = date[i].split('/')[0]
  y = date[i].split('/')[1]
  z = date[i].split('/')[2]
  month.append(x)
  day.append(y)
  year.append(z)

###ใส่ดาต้าใหม่

In [23]:
data_test_nonclane=data

In [24]:
data_test_nonclane['date']=date
data_test_nonclane['day']=day
data_test_nonclane['month']=month
data_test_nonclane['year']=year

In [25]:
#set(data_test_nonclane['day'])

###เเปลงให้ค่าเดือนอยู่ในหน่วยเดียวกัน

In [26]:
#data_test_nonclane[(data_test_nonclane['month']=='01')|(data_test_nonclane['month']=='02')|(data_test_nonclane['month']=='03')|(data_test_nonclane['month']=='04')|(data_test_nonclane['month']=='05')|(data_test_nonclane['month']=='06')|
                   #(data_test_nonclane['month']=='07')|(data_test_nonclane['month']=='08')|(data_test_nonclane['month']=='09')]

แปลงเดือน

In [27]:
data_test=data_test_nonclane.replace(to_replace=["1", "2",'3','4','5','6','7','8','9','10','11','12','01','02','03','04','05','06','07','08','09'],value=[1, 2,3,4,5,6,7,8,9,10,11,12,1,2,3,4,5,6,7,8,9])

In [28]:
set(data_test['month'])

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}

แปลงวัน

In [29]:
x=[]
for i in range(1,32):
  x.append(i)
print(x)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]


In [30]:
data_test=data_test_nonclane.replace(['01', '02', '03', '04', '05', '06', '07', '08', '09', '10','11',
                                      '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31'],x)

In [31]:
#set(data_test['day'])

###เริ่มทำข้อ 1


In [32]:
data_C_and_N = data_test[data_test['month']==12]

In [33]:
#!pip install apyori
#from apyori import apriori

In [34]:
transactionsC_and_N = data_C_and_N.groupby('InvoiceNo')['StockCode'].apply(list)

In [35]:
frequentItemsets = list(apriori(transactionsC_and_N,min_support = 0.05))
#frequentItemsets

In [36]:
min_sup = 0.04

frequentItemsetsC_and_N = list(apriori(transactionsC_and_N,min_support = min_sup))
for i in frequentItemsetsC_and_N:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)])


สรุปได้ดังนั้น
* RelationRecord(items=frozenset({'22910', '22086'}), support=0.047697368421052634)
  * ordered_statistics=OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.047697368421052634, lift=1.0)
  * OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5160142348754448, lift=7.195794834960331)
  * OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.6651376146788991, lift=7.195794834960332)


##2.ของประเทศในทวีป Europe

###เเบ่งทวีป

In [37]:
set(data['Country'])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

ทวีปยุโรป

In [38]:
UP_country = ['Austria','Belgium','Cyprus','Czech Republic','Denmark','EIRE','European Community','Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands','Norway','Poland','Portugal','Spain','Sweden','United Kingdom']

ทวีปเอเชีย

In [39]:
Asia_country = ['Hong Kong','Bahrain','Israel','Japan','Lebanon','Saudi Arabia','Singapore','United Arab Emirates']

In [40]:
#a0=data_test[data_test['Country']==UP_country[0]]

### หาข้อมูลทวีปยุโรป

In [41]:
EU_data = data_test[data_test['Country']==""]
for i in range(len(UP_country)):
  x = data_test[data_test['Country']==UP_country[i]]
  EU_data=pd.concat([x,EU_data])

In [42]:
len(set(EU_data['Country']))

22

###หาข้อ 2

In [43]:
transactionsEU = EU_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [44]:
frequentItemsets = list(apriori(transactionsEU,min_support = 0.05))
#frequentItemsets

In [45]:
min_sup = 0.03

frequentItemsetsEU= list(apriori(transactionsEU,min_support = min_sup))
for i in frequentItemsetsEU:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03243053387449266, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03243053387449266, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6772616136919316, lift=8.201394890946151), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39272211720226846, lift=8.201394890946151)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.030049953168904152, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.030049953168904152, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7389635316698656, lift=17.213819577735123), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.21381957773

##3.ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 jan - 15 Feb)

###เเยกทวีปเอเชีย

In [46]:
Asia_data = data_test[data_test['Country']==""]
for i in range(len(Asia_country)):
  x = data_test[data_test['Country']==Asia_country[i]]
  Asia_data=pd.concat([x,Asia_data])

In [47]:
len(set(Asia_data['Country']))

8

###หาข้อมูลช่วงตรุษจีน

In [48]:
M1_Asia_data=Asia_data[(Asia_data['month']==1)]

In [49]:
D1_Asia_data=M1_Asia_data[(M1_Asia_data['day']>=15)]

In [50]:
#D1_Asia_data

In [51]:
M1_Asia_data.head()

InvoiceNo StockCode                  Description  Quantity  \
47163   C540392     21155        RED RETROSPOT PEG BAG        -1   
47164   C540392     22244   3 HOOK HANGER MAGIC GARDEN        -1   
47165   C540392     22088  PAPER BUNTING COLOURED LACE       -40   
47166   C540392     22567      20 DOLLY PEGS RETROSPOT        -3   
50791    540556     21397       BLUE POLKADOT EGG CUP         12   

            InvoiceDate  UnitPrice  CustomerID Country        date  day month  \
47163  01/06/2011 18:06       2.10     12753.0   Japan  01/06/2011    6     1   
47164  01/06/2011 18:06       1.95     12753.0   Japan  01/06/2011    6     1   
47165  01/06/2011 18:06       2.55     12753.0   Japan  01/06/2011    6     1   
47166  01/06/2011 18:06       1.25     12753.0   Japan  01/06/2011    6     1   
50791  01/10/2011 09:58       1.25         NaN  Israel  01/10/2011   10     1   

       year  
47163  2011  
47164  2011  
47165  2011  
47166  2011  
50791  2011

In [52]:
M2_Asia_data=Asia_data[(Asia_data['month']==2)]

In [53]:
D2_Asia_data=M2_Asia_data[(M2_Asia_data['day']<=15)]

In [54]:
#D2_Asia_data

###เอาสองตารางรวมกัน

In [55]:
use_Asia_data=pd.concat([D1_Asia_data,D2_Asia_data])

###หาข้อ 3

In [112]:
transactionsAsia = D2_Asia_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [92]:
transactionsAsia

InvoiceNo
543179    [22407, 22406, 22405, 21745, 22496, 22495, 224...
543518    [22759, 22662, 22661, 21717, 22957, 22861, 222...
Name: StockCode, dtype: object

In [160]:
min_sup = 0.5
frequentItemsets_A= list(apriori(transactionsAsia,min_support = min_sup))
for i in frequentItemsets_A:
    if i[0].__len__()>1:
        print('yes')
        print(i)

KeyboardInterrupt: 

##4.ของช่วงเวลา Christmas & New Year ในทวีป Europe(1 Dec - 31 Dec)

In [82]:
C_N_EU_data=EU_data[EU_data['month']==12]

In [83]:
transactionsC_N_EU = C_N_EU_data.groupby('InvoiceNo')['StockCode'].apply(list)

In [94]:
transactionsC_N_EU

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 3022, dtype: object

In [88]:
min_sup = 0.04
frequentItemsetsC_N_EU= list(apriori(transactionsC_N_EU,min_support = min_sup))
for i in frequentItemsetsC_N_EU:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22910', '22086'}), support=0.04765056254136334, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22910', '22086'}), confidence=0.04765056254136334, lift=1.0), OrderedStatistic(items_base=frozenset({'22086'}), items_add=frozenset({'22910'}), confidence=0.5142857142857143, lift=7.162080315997367), OrderedStatistic(items_base=frozenset({'22910'}), items_add=frozenset({'22086'}), confidence=0.663594470046083, lift=7.162080315997367)])


##5.ของวันวาเลนไทน์ (14 Feb)

In [77]:
M14_data = data_test[(data_test['month']==2)]

In [78]:
va_data = M14_data[(M14_data['day']==14)]

In [79]:
M14_data[(M14_data['day']==14)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country, date, day, month, year]
Index: []

In [80]:
M14_data.head()

InvoiceNo StockCode                        Description  Quantity  \
77628    542776     21671       RED SPOT CERAMIC DRAWER KNOB        48   
77629    542776     21668     RED STRIPE CERAMIC DRAWER KNOB        12   
77630    542776     21670      BLUE SPOT CERAMIC DRAWER KNOB        12   
77631    542776    17013D       ORIGAMI ROSE INCENSE IN TUBE        96   
77632    542776     17096  ASSORTED LAQUERED INCENSE HOLDERS        48   

            InvoiceDate  UnitPrice  CustomerID         Country        date  \
77628  02/01/2011 08:23       1.25     15240.0  United Kingdom  02/01/2011   
77629  02/01/2011 08:23       1.25     15240.0  United Kingdom  02/01/2011   
77630  02/01/2011 08:23       1.25     15240.0  United Kingdom  02/01/2011   
77631  02/01/2011 08:23       0.42     15240.0  United Kingdom  02/01/2011   
77632  02/01/2011 08:23       0.21     15240.0  United Kingdom  02/01/2011   

       day month  year  
77628    1     2  2011  
77629    1     2  2011  
77630    1     2  2011  
77631    1     2  2011  
77632    1     2  2011

##6.จัดกลุ่มตามประเทศ+เดิอน

In [130]:
transactions_Mon_Con = data_test.groupby(['Country', 'month'])['StockCode'].apply(list)

In [133]:
transactions_Mon_Con

Country      month
Australia    1        [22078, 22079, 22080, 22077, 22505, 22516, 225...
             2        [84978, 22960, 47590B, 47566, 22087, 22720, 22...
             3        [22978, 22980, 22993, 22654, 22968, 22505, 225...
             4        [POST, 22232, 22993, 22449, 20726, 20725, 2072...
             5        [22699, 22859, 22730, 84663A, 23206, 48129, 23...
                                            ...                        
Unspecified  11       [22326, 20725, 22628, 22631, 84997D, 23254, 23...
             5        [23090, 47021G, 79030D, 84877A, 22178, 84913B,...
             6        [23122, 23119, 23108, 23240, 23237, 23239, 232...
             7        [22960, 16161P, 21499, 21500, 21503, 22047, 22...
             8        [23236, 21210, 23313, 23318, 22952, 22950, 232...
Name: StockCode, Length: 419, dtype: object

In [138]:
min_sup = 0.29
frequentItemsets_Mon_Con= list(apriori(transactions_Mon_Con,min_support = min_sup))
for i in frequentItemsets_Mon_Con:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'POST', '22326'}), support=0.29594272076372313, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'POST', '22326'}), confidence=0.29594272076372313, lift=1.0), OrderedStatistic(items_base=frozenset({'22326'}), items_add=frozenset({'POST'}), confidence=0.8435374149659863, lift=1.497636342672662), OrderedStatistic(items_base=frozenset({'POST'}), items_add=frozenset({'22326'}), confidence=0.5254237288135593, lift=1.497636342672662)])


#quiz 22386 เเละ 85099B คือสินค้าอะไร

In [ ]:
set(data[data['StockCode']=='22386']['Description'])

In [ ]:
set(data[data['StockCode']=='85099B']['Description'])

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [ ]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

In [ ]:
data[data['InvoiceNo']=='562950']